### PR TITLE
crossterm support

### LINKED ISSUE #283: Switch to Crossterm TUI backend
tui has switched to crossterm as the default, instead of termion

The reasoning given is here: https://github.com/fdehau/tui-rs/pull/553

I think it's pretty reasonable, and would probably make sense for us to switch too. Especially for the extra platform support :)

### PR DESCRIPTION
**Pull request by conradludgate**
For #282, fixes #283.

Todo

- [x] Fix failure to launch on 'up' keybinding
- [x] Fix up/down keys in the TUI (ctrl-n/p work fine)

### CODE REVIEW

**Comment by conradludgate** [comment]
A curiosity, I've decided to try partial upgrade (tui 0.18 + termion) and it's still eating my arrow key inputs, but otherwise still responsive. So I'm not sure what's going on there

**Comment by conradludgate** [comment]
> Do you know what mode your terminal is in? Different modes can send different escape sequences, and it could be listening for the wrong ones

Yeah, I was debugging and termion/crossterm never get the up/down. Which made me try again with our current version.

When running `atuin search -i` manually, it exhibits this behaviour.
When running from the keybinding, it is fine 🚀

**Comment by ellie** [comment]
> When running from the keybinding, it is fine 🚀

Yup I had this when I was writing the initial implementation. Some applications can switch terminal mode and then don't switch it back, leaving other programs in a broken state

Terminal stuff has a lot of baggage 😂

**Comment by conradludgate** [comment]
I'm rebasing this branch now, I'll let you know

**Comment by conradludgate** [comment]
https://github.com/crossterm-rs/crossterm/issues/396 :/

**Comment by conradludgate** [comment]
Ok, final conclusion since I've spent a while debugging this now. It's a combination of quite a few problems:

1. On ZSH, in our hook, stdin is not a tty. This is different to bash and fish
2. On macos, kqueue can't poll from /dev/tty
3. crossterm uses mio to poll from the tty, which uses kqueue on macos.

mio said they won't fix this specific behaviour, tbh I agree: https://github.com/tokio-rs/mio/issues/1377

There's an open PR for crossterm to manually use `select()` for /dev/tty: https://github.com/crossterm-rs/crossterm/pull/711

So far, this patch fixed it for me

File: src/command/client/search/interactive.rs
Comment (conradludgate) [inline src/command/client/search/interactive.rs:403]: moved the event process to after the render, this makes the first draw instant

File: src/command/client/search/interactive.rs
Comment (conradludgate) [inline src/command/client/search/interactive.rs:394]: poll(0) will ensure there's an event without blocking. This mimics the `try_next()` thing we did before 👌

**Comment by ellie** [comment]
@conradludgate looks like there's been some motion on that crossterm PR, and now this has been opened

https://github.com/crossterm-rs/crossterm/pull/735

**Comment by conradludgate** [comment]
Yep. I've been following it. Hopefully it gets merged at some point

**Comment by conradludgate** [comment]
Just waiting for a new crossterm release after the merge ⏳

**Comment by pdecat** [comment]
Crossterm 0.26 is out 🎉 
https://github.com/crossterm-rs/crossterm/releases/tag/0.26

**Comment by conradludgate** [comment]
I have updated to 0.26 crossterm and started vendoring a minimal subset of tui (now that it's maintenance is in question). It can be trimmed down further, but I don't want to go too minimal incase there are features we end up using.

File: src/tui/README.md
Comment (ellie) [inline src/tui/README.md:1]: Makes sense that we are vendoring this, but it would be nice to be able to remove it someday in The Future™️

**Review by ellie — APPROVED** [review]
LGTM!! Super excited we are finally ready to do this hahaha, thank you for working on this for so long.

File: src/tui/README.md
Comment (conradludgate) [reply src/tui/README.md:1]: Yeah, slowly we can either find a replacement or slim it down as much as possible to make the API as good for us as possible

File: src/tui/terminal.rs
Comment (pdecat) [inline src/tui/terminal.rs:8]: Hi @conradludgate @ellie,

I started to try and rebase my crossterm/inline PR https://github.com/ellie/atuin/pull/648, and figured that the version of tui-rs that was copied here is missing the bits needed from https://github.com/fdehau/tui-rs/pull/552/files#diff-554109cedfab40c7397c98d70e7a7449266358293b29ad5aa8db2f79e958122bR11 to make inline viewport possible.

Should I try and backport https://github.com/fdehau/tui-rs/pull/552 on this local copy?

Any reason why git history from upstream was not preserved using git subtree?

Also, wouldn't maintaining a fork of tui-rs make things easier, at least until it is not needed anymore at all?

Comment (conradludgate) [reply src/tui/terminal.rs:8]: A full fork WG is already in the works and I'm taking part as a stakeholder. It seems we won't be vendoring it for long.

I'm not an expert in all the git features so I didn't think to use subtree. It sounds like it would be awkward though because we want it as a module and not a crate

Comment (pdecat) [reply src/tui/terminal.rs:8]: Ok, I guess should I copy the parts needed from the upstream inline PR then.

**Review by ellie — APPROVED** [review]
Ok let's do it 🚀

**Pull request merged**
